anvil/src/shell/buffer.rs

```diff
@@ -0,0 +1,35 @@
+use smithay::{
+    reexports::{
+        calloop::Interest,
+        wayland_server::{
+            protocol::{wl_buffer::WlBuffer, wl_surface::WlSurface},
+            Resource,
+        },
+    },
+    wayland::{
+        buffer::BufferHandler,
+        compositor::{add_blocker, CompositorHandler},
+        dmabuf::get_dmabuf,
+    },
+};
+
+use crate::{state::Backend, AnvilState};
+
+impl<BackendData: Backend> BufferHandler for AnvilState<BackendData> {
+    fn buffer_attached(&mut self, buffer: &WlBuffer, surface: &WlSurface) {
+        if let Ok(dmabuf) = get_dmabuf(buffer) {
+            if let Ok((blocker, source)) = dmabuf.generate_blocker(Interest::READ) {
+                let client = surface.client().unwrap();
+                let res = self.handle.insert_source(source, move |_, _, data| {
+                    data.state
+                        .client_compositor_state(&client)
+                        .blocker_cleared(&mut data.state, &data.display.handle());
+                    Ok(())
+                });
+                if res.is_ok() {
+                    add_blocker(surface, blocker);
+                }
+            }
+        }
```

Comment:
How does that handle clients that attach multiple buffers during a single commit? I guess that would add blockers for buffers that are no longer in use?

What would happen if a client attaches and commits buffers faster then the render loop in the compositor? Like for example `glmar2-es2-wayland` which does not wait for frame callbacks. Wouldn't that also insert blockers for buffers that are potentially already released. Could that stall updates because the buffer is already in use by the client? If the client destroys that buffers, which can happen in case of glmark, it might not be an issue, but this is not guaranteed. iirc this highly depends on how `mesa` handles the buffers.